contracts/eosio.msig/src/eosio.msig.cpp

```diff
@@ -144,46 +185,34 @@ void multisig::exec( name proposer, name proposal_name, name executer ) {
 
    proposals proptable( get_self(), proposer.value );
    auto& prop = proptable.get( proposal_name.value, "proposal not found" );
-   transaction_header trx_header;
-   datastream<const char*> ds( prop.packed_transaction.data(), prop.packed_transaction.size() );
-   ds >> trx_header;
+   transaction_header trx_header = get_trx_header(prop.packed_transaction.data(), prop.packed_transaction.size());
    check( trx_header.expiration >= eosio::time_point_sec(current_time_point()), "transaction expired" );
 
-   approvals apptable( get_self(), proposer.value );
-   auto apps_it = apptable.find( proposal_name.value );
-   std::vector<permission_level> approvals;
-   invalidations inv_table( get_self(), get_self().value );
-   if ( apps_it != apptable.end() ) {
-      approvals.reserve( apps_it->provided_approvals.size() );
-      for ( auto& p : apps_it->provided_approvals ) {
-         auto it = inv_table.find( p.level.actor.value );
-         if ( it == inv_table.end() || it->last_invalidation_time < p.time ) {
-            approvals.push_back(p.level);
-         }
-      }
-      apptable.erase(apps_it);
+   auto table_op = [](auto&& table, auto&& table_iter) { table.erase(table_iter); };
+   bool ok = approvals_satisfy_trx_authorization(proposer, proposal_name, prop.packed_transaction, table_op);
+   check( ok, "transaction authorization failed" );
+
+   if ( prop.earliest_exec_time.has_value() && prop.earliest_exec_time.value().has_value() ) {
+      check( prop.earliest_exec_time.value().value() <= current_time_point(), "too early to execute" );
    } else {
-      old_approvals old_apptable( get_self(), proposer.value );
-      auto& apps = old_apptable.get( proposal_name.value, "proposal not found" );
-      for ( auto& level : apps.provided_approvals ) {
-         auto it = inv_table.find( level.actor.value );
-         if ( it == inv_table.end() ) {
-            approvals.push_back( level );
-         }
-      }
-      old_apptable.erase(apps);
+      check( trx_header.delay_sec.value == 0, "old proposals are not allowed to have non-zero `delay_sec`; cancel and retry" );
    }
-   auto packed_provided_approvals = pack(approvals);
-   auto res =  check_transaction_authorization(
-                  prop.packed_transaction.data(), prop.packed_transaction.size(),
-                  (const char*)0, 0,
-                  packed_provided_approvals.data(), packed_provided_approvals.size()
-               );
 
-   check( res > 0, "transaction authorization failed" );
+   datastream<const char*> ds = {prop.packed_transaction.data(), prop.packed_transaction.size()};
+   transaction_header _;
+   std::vector<action> context_free_actions;
+   std::vector<action> actions;
+   ds >> _;
+   ds >> context_free_actions;
+   ds >> actions;
+
+   for (const auto& act : context_free_actions) {
```

Comment:
Do not send the context free actions. Instead assert that there are no context free actions. In fact it would be a good idea to assert that at the transaction proposal stage as well.

Reply:
After doing that, add a test case demonstrating that the user is not allowed to propose a transaction with context free actions.

Reply:
Previously linked: https://github.com/EOSIO/eosio.contracts/blob/deprecate-deferred3/contracts/eosio.msig/src/eosio.msig.cpp#L188-L196
https://github.com/EOSIO/eosio.contracts/blob/deprecate-deferred3/contracts/eosio.msig/src/eosio.msig.cpp#L31-L36
https://github.com/EOSIO/eosio.contracts/blob/deprecate-deferred3/contracts/eosio.msig/src/eosio.msig.cpp#L208-L210
https://github.com/EOSIO/eosio.contracts/blob/deprecate-deferred3/tests/eosio.msig_tests.cpp#L1252-L1266
https://github.com/EOSIO/eosio.contracts/blob/deprecate-deferred3/tests/eosio.wrap_tests.cpp#L393-L407